keps/sig-cli/2953-kustomize-plugin-graduation/README.md

```diff
@@ -0,0 +1,764 @@
+
+**Plugin provider trust model:** Currently, the user assumes the risk associated with running the third party code by using the `--enable-alpha-plugins` flag. Starklark and exec plugins are not sandboxed in any way, and require additional `--enable-exec` /  `--enable-star` flags to invoke. Container-based plugins are run without disk or network access by default, but if a plugin requires it, the user must enable it both in the plugin config object (which they may not have written themselves) and on the command line via the `--mount`, `--network` and `--network-name` flags. The [Catalog KEP] proposes a way to trust a published set of plugins as well as a way for Kustomize to validate the integrity of that the providers it fetches.
+
+### Goals
+
+1. Kustomize's extension story MUST be consistent between standalone Kustomize and kustomize-in-kubectl. For example, if any flags are used to enable or configure Kustomize, the usage of those flags must be identical in both the standalone and embedded distributions.
+1. Plugin configuration MUST be fully declarative.
+1. Plugin usage MUST NOT require out-of-band imperative installation steps. 
+1. Kustomize SHOULD support declarative configuration of a set of trusted plugins.  (See the [Catalog KEP])
+1. Kustomize SHOULD provide an API that enables plugins to be orchestrated on equal footing with built-in transformers, including with respect to ordering. (See the [Composition KEP])
+1. Kustomize SHOULD provide an API that facilitates sane layering in plugin-centric workflows where the plugin configuration itself must be manipulated. (See the [Composition KEP])
+1. As a measure of plugin UX, Kustomize itself SHOULD be able to convert existing transformers (e.g. the Helm transformer) to "official" plugins without significantly degrading the user experience.
+
+### Non-Goals
+
+1. Make any change to the Configuration Functions Specification, a low-level spec shared with other orchestrators such as kpt. No such changes are required by this KEP.
+1. Create a catalog of "official" plugins. Although such a thing would be a natural extension to this KEP, it is out of scope.
+1. Convert currently compiled in transformers to plugins. Although this KEP should make this possible, the work to do so is out of scope.
+
+## Proposal
+
+This KEP proposes to deprecate the legacy style plugins in favour of the KRM Functions style, with some refinements. The KRM Functions style plugin mechanism is the most consistent with Kustomize's philosophy and most readily able to be adapted to meet the requirements outlined in the Goals section. It is also based on an open standard that has been adopted by at least one other tool (kpt), increasing the value of plugins built in this way.
+
+This KEP proposes implementing both the [Composition KEP] and the [Catalog KEP] as a prerequisite for plugin GA in Kustomize. Those two KEPs solve key discovery, distribution and UX problems described in the goals section.
```

Comment:
I think we need to be mindful of a use case where someone just wants to have a plugin that they write and don't want to create a catalog.  Here is a scenario:
1. User creates a plugin that ensures that all instances are capped at 10 (maybe it's a billing policy)
2. User has just that plugin and is in charge of their CI/CD images so it gets into the build machine image for CloudBuild or CircleCI along with kustomize.
3. User wants to test the results of the kustomization locally during development without having to create a catalog
4. User wants to push out their custom CICD pipeline without having to stand up additional servers or services.

I am weary of introducing additional infrastructure requirements on someone like that.  One of the key selling points of kustomize is that it managed to keep simple things simple (namespace for instance).  I'd like to apply the same philosophy here if possible.

Reply:
This is a very similar scenario to user story 1 below. Did you have a look at that? There are no additional infrastructure requirements, and in fact the ad-hoc plugin would be distributed as part of the Kustomization bundle, giving a more consistent, lower-effort experience across local and remote environments. A catalog does need to be generated, but the catalog generation is a simple one-time command to run in this proposal. Each invoker does need to trust the catalog, but there's no getting around that since running plugins without explicit user acknowledgement in the invocation is unacceptable for security.